File: Day #1061/solution.cpp
// Shortest path on a boolean grid (true=wall) via BFS in 4 directions.
// Time: O(M*N), Space: O(M*N). Returns -1 (null) if unreachable.
#include <bits/stdc++.h>
using namespace std;

int shortestPath(const vector<vector<bool>>& board, pair<int,int> start, pair<int,int> end) {
    int m = board.size(), n = board[0].size();
    if (board[start.first][start.second] || board[end.first][end.second]) return -1;
    vector<vector<bool>> seen(m, vector<bool>(n, false));
    queue<pair<int,int>> q;
    q.push(start);
    seen[start.first][start.second] = true;
    int steps = 0;
    int dr[] = {-1, 1, 0, 0}, dc[] = {0, 0, -1, 1};
    while (!q.empty()) {
        int sz = q.size();
        for (int i = 0; i < sz; ++i) {
            int r = q.front().first, c = q.front().second; q.pop();
            if (r == end.first && c == end.second) return steps;
            for (int d = 0; d < 4; ++d) {
                int nr = r + dr[d], nc = c + dc[d];
                if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
                if (seen[nr][nc] || board[nr][nc]) continue;
                seen[nr][nc] = true;
                q.push(make_pair(nr, nc));
            }
        }
        ++steps;
    }
    return -1;
}

int main() {
    vector<vector<bool>> board = {
        {false, false, false, false},
        {true,  true,  false, true },
        {false, false, false, false},
        {false, false, false, false}
    };
    int res = shortestPath(board, {3, 0}, {0, 0});
    if (res == -1) cout << "null\n";
    else cout << res << "\n";
    return 0;
}

File: Day #1061/solution.py
# Day 1061: Shortest path on a boolean grid (True=wall) via BFS in 4 directions.
# Time: O(M*N), Space: O(M*N). Returns None if unreachable.
from collections import deque


def shortest_path(board, start, end):
    m, n = len(board), len(board[0])
    if board[start[0]][start[1]] or board[end[0]][end[1]]:
        return None
    seen = [[False] * n for _ in range(m)]
    q = deque([(start[0], start[1], 0)])
    seen[start[0]][start[1]] = True
    while q:
        r, c, steps = q.popleft()
        if (r, c) == end:
            return steps
        for dr, dc in ((-1, 0), (1, 0), (0, -1), (0, 1)):
            nr, nc = r + dr, c + dc
            if 0 <= nr < m and 0 <= nc < n and not seen[nr][nc] and not board[nr][nc]:
                seen[nr][nc] = True
                q.append((nr, nc, steps + 1))
    return None


if __name__ == "__main__":
    f, t = False, True
    board = [
        [f, f, f, f],
        [t, t, f, t],
        [f, f, f, f],
        [f, f, f, f],
    ]
    print(shortest_path(board, (3, 0), (0, 0)))
